content/blog/elm-warrior-walkthrough/index.md

```diff
@@ -0,0 +1,818 @@
+
+## SPOILERS AHEAD!
+## SPOILERS AHEAD!!
+## SPOILERS AHEAD!!!
+
+Now that my conscience is clear, I'll start by copying the exposed values
+and functions from each top-level module and pasting them all together. No
+docs, no type signatures, nothing, just their **names**. As [the sage once
+said](https://cs.fit.edu/~ryan/rose.html):
+
+> Stat rosa pristina nomine, nomina nuda tenemus.
+
+Which roughly translates as:
+
+> The rose of old remains only in its name. Only naked names we possess.
+
+Let's get moving, shall we?
+
+```elm
+module Warrior exposing
+    ( Warrior, Action(..)
+    , id, position
+    , health, maxHealth, healingPotential
+    , attackDamage, inventory
+    )
+
+module Warrior.Coordinate exposing
+    ( Coordinate
+    , toString
+    )
+
+module Warrior.Direction exposing
+    ( Direction(..)
+    , all, toString
+    )
+
+module Warrior.History exposing
+    ( History
+    , roundsPlayed, previousStates, previousActions
+    )
+
+module Warrior.Item exposing
+    ( Item(..)
+    , toString
+    )
+
+module Warrior.Map exposing
+    ( Map
+    , look, lookDown
+    )
+
+module Warrior.Program exposing
+    ( TurnFunction
+    , Model, Msg
+    , Config, program
+    )
+```
+
+At this point we clearly don't understand what is going on with the code,
+but just by reading this listing a couple of times we are getting familiar
+with the _names_ of this particular domain. Just let it soak in.
+
+Now I'm getting more curious and will take a look at these exposed data
+constructors: `Direction`, `Action`, and `Item`. Whenever a module exposes
+the inner contents of a type, that's an invitation to the world to depend
+on those values, so I'm sure it's going to be useful to take a peek.
+
+```elm
+type Direction
+    = Left
+    | Right
+    | Up
+    | Down
+
+type Action
+    = Wait
+    | Heal
+    | Pickup
+    | Move Direction
+    | Attack Direction
+
+type Item
+    = Sword
+    | Potion
+```
+
+There is a lot of information packed in these simple names, so don't worry
+if you feel a bit overwhelmed. Instead, let's try something practical and
+look at the first level.
+
+## Level I
+
+Let's start [here](https://ellie-app.com/989KZm4BrVMa1).
+
+The source on the left hand side reads like this:
+
+```elm
+main : Program () Program.Model Program.Msg
+main =
+    Program.program
+        { maps = Maps.all
+        , players = [ ( "Player", takeTurn ) ]
+        , msPerTurn = 500
+        , progressionFunction = Progression.reachExitPoint
+        }
+
+
+takeTurn : Warrior -> Map -> History -> Warrior.Action
+takeTurn warrior map history =
+    Warrior.Wait
+```
+
+We can see on the right hand side that every 500 ms a `Player waits`
+message is being printed. So instead of simply waiting, we have to make our
+warrior move to the right. Let's import `Direction` and change the
+`takeTurn` function:
+
+```elm
+takeTurn : Warrior -> Map -> History -> Warrior.Action
+takeTurn warrior map history =
+    Warrior.Move Right
```

Comment:
At this point I didn't know you were going to provide links to all intermediate steps and made this change myself, planning to follow along. When doing that, I got a compilation error because of missing imports/exposed values.

That's not terrible, but might discourage some folks thinking that this is not polished enough or that they'll have less hand-holding than what you actually provide. Maybe a heads up that that will happen and/or that you will provide a working link to the solution would be nice.

Reply:
Done, thank you!